### PR TITLE
KAFKA-7514: Add threads to ConsumeBenchWorker

### PR DESCRIPTION
**Pull request by stanislavkozlovski**
This PR adds a new ConsumeBenchSpec field - "consumerCount". "consumerCount" will be spawned over "consumerCount" threads in the ConsumeBenchWorker.
Since "consumerCount" will be 1 by default, these changes are backwards compatible

It's now questionable how existing fields such as "targetMessagesPerSec", "maxMessages", "consumerGroup" and "activeTopics" should work.

With "activeTopics", we need to decide whether they should be split over the consumers or not.
I see 4 cases which I believe we should address like this:

1. Random group, subscribe to topics - N unique groups all subscribed to all topics
2. Specifed group, subscribe to topics - 1 group subscribed to all topics. Consumers share workload.
3. Random groups, assign partitions - X groups all subscribed to all partitions
4. Specified group, assign partitions - Throw an exception. Only one consumer can read from a specific partition within the context of a consumer group. It is then unclear how and whether at all we should split the partitions across the consumers. At this phase, I believe it's best to not support this.

I believe "targetMessagesPerSec", "maxMessages" should account for each consumer individually. This would ease implementation by a ton, too.

I haven't written tests yet since I want to flesh out the design first. Any feedback is appreciated

### CODE REVIEW

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java [hunk 1]
```diff
@@ -182,9 +237,11 @@ public Void call() throws Exception {
             long bytesConsumed = 0;
             long startTimeMs = Time.SYSTEM.milliseconds();
             long startBatchMs = startTimeMs;
+            int maxMessages = spec.maxMessages();
             try {
-                while (messagesConsumed < spec.maxMessages()) {
-                    ConsumerRecords<byte[], byte[]> records = consumer.poll(Duration.ofMillis(50));
+                while (messagesConsumed < maxMessages) {
+
+                    ConsumerRecords<byte[], byte[]> records = consumer.poll();
```
Comment (stanislavkozlovski) [inline tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:239]: This raises an `InterruptException` and I can't figure out why. Is there something obvious that I'm missing?
```
[2018-11-05 17:47:05,859] WARN ConsumeRecords caught an exception:  (org.apache.kafka.trogdor.workload.ConsumeBenchWorker)
org.apache.kafka.common.errors.InterruptException: java.lang.InterruptedException
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.maybeThrowInterruptException(ConsumerNetworkClient.java:493)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:281)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:236)
	at org.apache.kafka.clients.consumer.KafkaConsumer.pollForFetches(KafkaConsumer.java:1243)
	at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1188)
	at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1164)
	at org.apache.kafka.trogdor.workload.ConsumeBenchWorker$Consumer.poll(ConsumeBenchWorker.java:503)
	at org.apache.kafka.trogdor.workload.ConsumeBenchWorker$ConsumeMessages.call(ConsumeBenchWorker.java:243)
	at org.apache.kafka.trogdor.workload.ConsumeBenchWorker$ConsumeMessages.call(ConsumeBenchWorker.java:198)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.lang.InterruptedException
	... 14 more
```

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java [hunk 2]
```diff
@@ -202,62 +259,140 @@ public Void call() throws Exception {
...
                         bytesConsumed += messageBytes;
+                        if (messagesConsumed >= maxMessages)
+                            break;
+
                         throttle.increment();
                     }
                     startBatchMs = Time.SYSTEM.milliseconds();
                 }
             } catch (Exception e) {
+                // TODO: Should we close task on consumer failure?
```
Comment (stanislavkozlovski) [inline tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java]: This is an open question - I'm not sure what's better - to abort the whole task if one consumer fails or continue running until the last one fails

Comment (cmccabe) [reply tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java]: Please abort the whole task if the consumer fails.  That's how the rest of the tests work.

In general, the consumer should never fail-- if it does, there is a serious problem.

**Comment by cmccabe** [comment]
Please let's use "thread count" instead of "consumer count."  We may eventually want to extend this job to be able to spawn multiple workers (if it isn't already).

I think the questions above about how max messages per sec, etc. work are resolved, right?  They apply on a per-thread basis.  We should add some JavaDoc specifying this.

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java (hunk 1 above)
Comment (cmccabe) [reply tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:239]: Most likely something somewhere else failed, and your thread was sent the interrupt exception as part of tearing down the worker.

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java (hunk 2 above)
Comment (stanislavkozlovski) [reply tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java]: Makes sense. I have logic implemented that does not fail the whole task is a `StatusUpdater` of one consumer fails - only when all `StatusUpdaters` fail does the task get aborted. Thoughts on that?

**Comment by stanislavkozlovski** [comment]
Thanks for the review @cmccabe. The questions on maxMessages and etc are resolves, yes. I have already added a JavaDoc explaining this in `ConsumeBenchSpec.java` - please take a look to see if they are adequate

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java (hunk 1 above)
Comment (stanislavkozlovski) [reply tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:239]: I am not sure about the cause of this since I haven't been able to reproduce it since. I had changed the parameters in `trogdor-run-consume-bench.sh`. Might have been something to do with old consumer group state after having re-ran it a couple of times

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java (hunk 2 above)
Comment (cmccabe) [reply tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java]: The status updaters should not fail, right?  Any failure should abort the worker.

Comment (stanislavkozlovski) [reply tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java]: Seems reasonable. I reverted my commit

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchSpec.java [hunk 3]
```diff
@@ -100,6 +108,7 @@ public ConsumeBenchSpec(@JsonProperty("startMs") long startMs,
                             @JsonProperty("consumerConf") Map<String, String> consumerConf,
                             @JsonProperty("commonClientConf") Map<String, String> commonClientConf,
                             @JsonProperty("adminClientConf") Map<String, String> adminClientConf,
+                            @JsonProperty("threadCount") Integer threadCount,
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchSpec.java]: Can we call this "threadsPerWorker" instead?  We will probably want to support multiple workers in the future.

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchSpec.java [hunk 4]
```diff
@@ -138,6 +148,11 @@ public int maxMessages() {
         return maxMessages;
     }
 
+    @JsonProperty
+    public int consumerCount() {
+        return consumerCount;
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchSpec.java]: We should make this consistent with the name of the varialbe in the spec

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchSpec.java [hunk 5]
```diff
@@ -61,6 +62,12 @@
  * #{@link org.apache.kafka.clients.consumer.KafkaConsumer#subscribe(Collection)}.
  * It will be assigned partitions dynamically from the consumer group.
  *
+ * This specification supports the spawning of multiple consumers in the single Trogdor worker agent.
+ * The "threadCount" field denotes how many consumers should be spawned for this spec.
+ * It is worth noting that the "targetMessagesPerSec", "maxMessages" and "activeTopics" fields apply for every consumer individually.
+ * If a consumer group is not specified, every consumer is assigned a different, random group. When specified, all consumers use the same group.
+ * Specifying partitions, a consumer group and multiple consumers will result in an #{@link ConfigException} and the task will abort.
+ *
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchSpec.java:73]: It would probably be good to have a separate paragraph to describe the group configuration.

We probably want to mention that if individual partitions are specified, only a single consumer is supported, and subscribe is not used.

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java [hunk 6]
```diff
@@ -77,9 +85,12 @@ public void start(Platform platform, WorkerStatusTracker status,
             throw new IllegalStateException("ConsumeBenchWorker is already running.");
         }
         log.info("{}: Activating ConsumeBenchWorker with {}", id, spec);
+        this.statusUpdater = new StatusUpdater();
         this.executor = Executors.newScheduledThreadPool(
-            2, ThreadUtils.createThreadFactory("ConsumeBenchWorkerThread%d", false));
-        this.status = status;
+            spec.consumerCount() + 2, // 1 thread for all the ConsumeStatusUpdater and 1 for the StatusUpdater
+            ThreadUtils.createThreadFactory("ConsumeBenchWorkerThread%d", false));
+        this.statusUpdaterFuture = executor.scheduleAtFixedRate(this.statusUpdater, 2, 1, TimeUnit.MINUTES);
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java]: Hmm.  Why change the initial delay from 1 minute to 2?

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java [hunk 7]
```diff
@@ -88,41 +99,78 @@ public void start(Platform platform, WorkerStatusTracker status,
...
+        private List<ConsumeMessages> consumeTasks() {
+            List<ConsumeMessages> tasks = new ArrayList<>();
+            String consumerGroup = consumerGroup();
+            int consumerCount = spec.consumerCount();
             Map<String, List<TopicPartition>> partitionsByTopic = spec.materializeTopics();
-            boolean toUseGroupPartitionAssignment = partitionsByTopic.values().isEmpty();
+            boolean toUseGroupPartitionAssignment = partitionsByTopic.values().stream().allMatch(List::isEmpty);
+
+            if (!toUseGroupPartitionAssignment && !toUseRandomConsumeGroup() && consumerCount > 1)
+                throw new ConfigException(String.format("Will not split partitions across consumers from the %s group", consumerGroup));
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java]: This exception seems a bit confusing.

Maybe something like "You may not specify an explicit partition assignment when using multiple consumers in the same group.  Please either leave the consumer group unset, or specify topics instead of partitions."

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java [hunk 8]
```diff
@@ -88,41 +99,78 @@ public void start(Platform platform, WorkerStatusTracker status,
...
+
+            String clientId = clientId(0);
+            consumer = consumer(consumerGroup, clientId);
+            if (!toUseGroupPartitionAssignment) {
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java]: Since we're handling both the case where this is true, and where it is false, we might as well test for (toUseGroupPartitionAssignment) rather than (!toUseGroupPartitionAssignment)

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java [hunk 9]
```diff
@@ -88,41 +99,78 @@ public void start(Platform platform, WorkerStatusTracker status,
...
+            return String.format("consumer.%s-%d", id, idx);
         }
 
-        private KafkaConsumer<byte[], byte[]> consumer(String consumerGroup) {
+        /**
+         * Creates a new KafkaConsumer instance
+         */
+        private ThreadSafeConsumer consumer(String consumerGroup, String idx) {
             Properties props = new Properties();
+            String clientId = String.format("consumer.%s-%s", id, idx);
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java]: This seems to duplicate the code in the `clientId(int idx)` function?

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java [hunk 10]
```diff
@@ -88,41 +99,78 @@ public void start(Platform platform, WorkerStatusTracker status,
...
+        private String consumerGroup() {
+            return toUseRandomConsumeGroup()
+                ? "consume-bench-" + UUID.randomUUID().toString()
+                : spec.consumerGroup();
         }
 
-        private String generateConsumerGroup() {
-            return "consume-bench-" + UUID.randomUUID().toString();
+        private boolean toUseRandomConsumeGroup() {
+            return spec.consumerGroup().equals(ConsumeBenchSpec.EMPTY_CONSUMER_GROUP);
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java]: You should be able to just say `spec.consumerGroup().isEmpty()`

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java [hunk 11]
```diff
@@ -151,29 +199,35 @@ private String generateConsumerGroup() {
...
-                this.consumer.subscribe(topics);
-            } else {
-                List<TopicPartition> partitions = topicPartitionsByTopic.values().stream()
-                    .flatMap(List::stream).collect(Collectors.toList());
-                log.info("Will consume from topic partitions {} via manual assignment.", partitions);
-                this.consumer.assign(partitions);
-            }
+                new ConsumeStatusUpdater(latencyHistogram, messageSizeHistogram, consumer), 1, 1, TimeUnit.MINUTES);
+            int perPeriod;
+            if (spec.targetMessagesPerSec() == 0)
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java]: Suggest using <= 0

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java [hunk 12]
```diff
@@ -202,6 +257,9 @@ public Void call() throws Exception {
                         latencyHistogram.add(elapsedBatchMs);
                         messageSizeHistogram.add(messageBytes);
                         bytesConsumed += messageBytes;
+                        if (messagesConsumed >= maxMessages)
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:257]: Why is this needed?  The loop condition is the same.

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java [hunk 13]
```diff
@@ -333,11 +452,86 @@ public void stop(Platform platform) throws Exception {
...
+        this.statusUpdater = null;
+        this.statusUpdaterFuture = null;
+        this.workerStatus = null;
         this.doneFuture = null;
     }
 
+    /**
+     * A thread-safe KafkaConsumer wrapper
+     */
+    private static class ThreadSafeConsumer {
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:464]: This is not needed.  Only one thread at a time should be accessing the consumer.

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java [hunk 14]
```diff
@@ -333,11 +452,86 @@ public void stop(Platform platform) throws Exception {
...
+        ThreadSafeConsumer(KafkaConsumer<byte[], byte[]> consumer, String clientId) {
+            this.consumer = consumer;
+            this.clientId = clientId;
+            this.consumerLock = new ReentrantLock();
+        }
+
+        ConsumerRecords<byte[], byte[]> poll() {
+            this.consumerLock.lock();
+            try {
+                return consumer.poll(Duration.ofMillis(50));
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:479]: 50 ms is way too short for a poll interval.  We don't want to use up so much CPU.  If there's nothing to consume it should wait for at least a minute.

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java (hunk 6 above)
Comment (stanislavkozlovski) [reply tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java]: Whoops, I initially thought this was seconds. My reasoning was that since there are more threads spawned and a rebalance happening, it would make sense to wait a bit more before exposing the status.
Anyway, reverting this back to 1 minute

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java (hunk 8 above)
Comment (stanislavkozlovski) [reply tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java]: True, that's more readable

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java (hunk 12 above)
Comment (stanislavkozlovski) [reply tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:257]: The default `max.poll.records` is 500. This would result in the consumer always consuming 500 at a minimum (provided the topic had the records) and not respect the cases where `maxMessages` was less than that

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java (hunk 13 above)
Comment (stanislavkozlovski) [reply tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:464]: The `ConsumeStatusUpdater`  calls `KafkaConsumer#assignment()` dynamically to get the latest assignments for a consumer.
My reasoning was that there might be edge cases where these assignments change during a benchmark run (e.g one consumer finishes early, starts late, etc).
That might not be needed in practice, though. WDYT?

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java (hunk 14 above)
Comment (stanislavkozlovski) [reply tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:479]: That's how it before but good point. If we remove the lock we should go with this

File: tests/kafkatest/services/trogdor/consume_bench_workload.py [hunk 15]
```diff
@@ -21,7 +21,7 @@
 class ConsumeBenchWorkloadSpec(TaskSpec):
     def __init__(self, start_ms, duration_ms, consumer_node, bootstrap_servers,
                  target_messages_per_sec, max_messages, active_topics,
-                 consumer_conf, common_client_conf, admin_client_conf, consumer_group=None):
+                 consumer_conf, common_client_conf, admin_client_conf, consumer_group=None, consumer_count=1):
```
Comment (cmccabe) [inline tests/kafkatest/services/trogdor/consume_bench_workload.py]: Should be "threadsPerWorker"

File: tests/kafkatest/tests/core/consume_bench_test.py [hunk 16]
```diff
@@ -120,13 +143,62 @@ def test_consume_group_bench(self):
                                                 consumer_conf={},
                                                 admin_client_conf={},
                                                 common_client_conf={},
+                                                consumer_count=2,
```
Comment (cmccabe) [inline tests/kafkatest/tests/core/consume_bench_test.py]: Same (should be threads_per_worker)

**Comment by cmccabe** [comment]
Thanks @stanislavkozlovski .  Looks good.  Should be ready to commit once the variable name is fixed in the python code.

**Comment by stanislavkozlovski** [comment]
@cmccabe the JDK 8 tests passed and the JDK11 failures look like unrelated flaky tests
